torch/csrc/jit/passes/inliner.cpp

```diff
@@ -0,0 +1,115 @@
+#include <torch/csrc/jit/passes/inliner.h>
+#include <torch/csrc/jit/script/module.h>
+#include <torch/csrc/jit/script/error_report.h>
+
+namespace torch {
+namespace jit {
+
+namespace prim {
+using namespace ::c10::prim;
+}
+
+class Inliner {
+ public:
+  Inliner() = default;
+
+  void removeFunctionConstants(Block* b)
+  {
+    for (auto it = b->nodes().begin(); it != b->nodes().end(); it++)
+    {
+      if (it->kind() == prim::Constant && it->output()->type()->isSubclass(TypeKind::FunctionType))
+      {
+        //Inliner inlined all possible function constants
+        AT_ASSERT(it->output()->uses().size() == 0);
+        it.destroyCurrent();
+      }
+      else
+      {
+        for (auto ib : it->blocks())
+        {
+            removeFunctionConstants(ib);
+        }
+      }
+    }
+  }
+
+  void unifyIfs(Block* b)
+  {
+    for (auto n : b->nodes())
+    {
+      for (auto ib : n->blocks())
+      {
+          unifyIfs(ib);
+      }
+
+      if (n->kind() == prim::If)
+      {
+        auto true_outputs = n->blocks().at(0)->outputs();
+        auto false_outputs = n->blocks().at(1)->outputs();
+        AT_ASSERT(true_outputs.size() == false_outputs.size());
+        AT_ASSERT(true_outputs.size() == n->outputs().size());
+        for (size_t i = 0; i < true_outputs.size(); i++)
+        {
+          auto unified = unifyTypes(true_outputs.at(i)->type(), false_outputs.at(i)->type());
+          if (!unified) {
+            throw script::ErrorReport()
+                << "if-expression's true branch has type " << true_outputs.at(i)->type()->str()
+                << " but false branch has type " << false_outputs.at(i)->type()->str();
+          }
+          n->outputs()[i]->setType(*unified);
+        }
+
+      }
+    }
+
+  }
+  void inlineCalls(Block* block) {
+    Node* cur = block->nodes().front();
+    Node* end = block->return_node();
+
+    while (cur != end) {
+      auto next = cur->next();
+      if (cur->kind() == prim::CallFunction) {
+        AT_ASSERT(cur->inputs().at(0)->node()->kind() == prim::Constant);
+        auto function_constant = cur->inputs().at(0)->node();
+        auto fun_type = function_constant->output()->type()->expect<FunctionType>();
+        auto graph = fun_type->function()->graph();
+
+        auto old_output = cur->outputs();
+        // slice function ptr value
+        auto inputs = cur->inputs().slice(1);
+        WithInsertPoint guard(next);
+        auto new_output =
+            inlineCallTo(*cur->owningGraph(), *graph.get(), inputs).at(0);
+        if (old_output.at(0)->hasUniqueName()) {
+          auto name = old_output.at(0)->uniqueName();
+          new_output->setUniqueName(name);
+        }
+
+        old_output.at(0)->replaceAllUsesWith(new_output);
+        next = cur->next();
+        cur->destroy();
+      } else {
+        if (cur->hasAttribute(attr::Subgraph)) {
```

Comment:
Optimization of subgraphs should be handled by the thing inserting the subgraph. Most of our passes do not descend into subgraphs, which is why this code needs to keep modifying passes. I don't want to start mixing typical behavior and have some passes look at subgraphs but others that do not.

Reply:
I think that inliner is a **very** special case. It is more similar to `compiler.cpp` rather than a regular optimization. Let me check if making `compiler.cpp` calling `Inline` is good enough. I think it was still breaking some tests if `Inliner` didn't expect subgraphs.